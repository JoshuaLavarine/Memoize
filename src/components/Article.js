import React, { Component } from "react";
import Quiz from './Quiz';

class Article extends Component {
  constructor() {
    super();
    this.state = {
      displayQuiz: false,
      incorrectAnswers: {
        1: [1, 2, 3, 4, 5],
        2: [1, 2, 3, 4, 5],
        3: [1, 2, 3, 4, 5],
        4: [1, 2, 3, 4, 5],
        5: [1, 2, 3, 4, 5]
      }
    }
  }

  componentDidMount() {
    this.hydrateStateWithLocalStorage();
  }

  hydrateStateWithLocalStorage() {
    // for all items in state
    for (let key in this.state) {
      // if the key exists in localStorage
      if (localStorage.hasOwnProperty(key)) {
        // get the key's value from localStorage
        let value = localStorage.getItem(key);

        // parse the localStorage string and setState
        try {
          value = JSON.parse(value);
          this.setState({ [key]: value });
        } catch (e) {
          // handle empty string
          this.setState({ [key]: value });
        }
      }
    }
  }

  displayQuiz = () => {
    this.setState({
      displayQuiz: true
    })
  }

  displayPioneer = () => {
    let selectedPioneer = this.props.pioneers.find(pioneer => {
      return pioneer.id == this.props.currentPioneer
    })
    return ( 
      <img className="pioneer" src={selectedPioneer.img}></img>
    )
  }

  displayArticle = () => {
    let selectedPioneer = this.props.pioneers.find(pioneer => {
      return pioneer.id == this.props.currentPioneer
    })
    let allParagraphs = selectedPioneer.paragraphs.map(paragraph => {
      return <p>{paragraph}</p>
    })
    return allParagraphs
  }


  resetIncorrectAnswers = () => {
    let selectedPioneer = this.props.pioneers.find(pioneer => {
      return pioneer.id == this.props.currentPioneer
    })
    // console.log('conditional', this.state.incorrectAnswers[selectedPioneer.id].length)
    if (this.state.incorrectAnswers[selectedPioneer.id].length === 0) {
      console.log('resetthe state')
      this.setState(prevState => ({
        incorrectAnswers: {
            ...prevState.incorrectAnswers,
          [selectedPioneer.id]: [1, 2, 3, 4, 5]
        }
      }))
    }
    console.log('before LS', localStorage) 
    localStorage.setItem('incorrectAnswers', JSON.stringify(this.state.incorrectAnswers));
    console.log('after LS', localStorage)
  }

  displayQuizAndResetIncorrectAnswers = () => {
    this.resetIncorrectAnswers()
    this.displayQuiz()
  }

  render() {
    switch(this.state.displayQuiz) {
      case(true):
      return (
        <Quiz 
        currentPioneer={this.props.currentPioneer}
        pioneers={this.props.pioneers}
        selectPioneer = {this.props.selectPioneer}
        incorrectAnswers = {this.state.incorrectAnswers}
        />
      )
    default:
      return (
        <section className="complete-article">
          <article className='pioneer-img-article'>
            {this.displayPioneer()}
          </article>
          <article className='pioneer-article'>
            {this.displayArticle()}
          </article>
          <section className="button-container">
            <button className="button-back" onClick={this.props.selectPioneer} value={0}>Go Back</button>
            <button className="button-quiz" onClick={this.displayQuizAndResetIncorrectAnswers}>Go to Quiz</button>
          </section>
        </section>
      )
    }
  }
}

export default Article;