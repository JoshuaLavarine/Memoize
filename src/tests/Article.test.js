import React from "react";
import Article from "../Components/Article";
import { shallow } from "enzyme";
import { pioneers } from "./mockData";

describe("Article", () => {
	const mockSelectPioneer = jest.fn();
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
			<Article 
			currentPioneer={1}
			pioneers={pioneers}
			selectPioneer = {mockSelectPioneer}
		/>    
    );
  });
	
  it("should match the snapshot with all data passed in", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should have a proper default state", () => {
    expect(wrapper.state()).toEqual({ 
			displayQuiz: false,
      incorrectAnswers: {
        1: [1, 2, 3, 4, 5],
        2: [1, 2, 3, 4, 5],
        3: [1, 2, 3, 4, 5],
        4: [1, 2, 3, 4, 5],
        5: [1, 2, 3, 4, 5]
			}
		});
	});
	
	it.skip("should change the displayQuiz to true", () => {
		
		// expect(
	});

});