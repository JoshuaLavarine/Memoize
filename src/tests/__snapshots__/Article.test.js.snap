// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article should match the alternative snapshot with all data passed in 1`] = `
<section
  className="complete-article"
>
  <article
    className="pioneer-img-article"
  >
    <img
      className="pioneer"
      src="http://www.greatthoughtstreasury.com/sites/default/files/Jennifer-Pahlka1%5B1%5D.jpg"
    />
  </article>
  <article
    className="pioneer-article"
  >
    <p
      key="0"
    >
      Jennifer Pahlka (1969 - present day) is the founder and Executive Director of Code for America. She served as US Deputy Chief Technology Officer from June 2013 to June 2014 and helped found the United States Digital Service. Previously she had worked at CMP Media with various roles in the computer game industry. She was the co-chair and general manager of the Web 2.0 conferences.
    </p>
    <p
      key="1"
    >
      She was born in Port Deposit, Maryland, and raised in Austin, New Haven, and New York City. She is a graduate of the Bronx High School of Science and Yale University.
    </p>
    <p
      key="2"
    >
      Pahlka founded Code for America, a San Francisco-based non-profit organization that aims to make government more transparent and connected. According to the Washington Post Code for America is the technology world’s equivalent of the Peace Corps or Teach for America, offering an alternative to the old, broken path of government IT. In her 2012 TED Talk, Pahlka noted that we will not be able to reinvent government unless we also reinvent citizenship.
    </p>
    <p
      key="3"
    >
      In May 2013 Pahlka announced she was temporarily taking the position of deputy chief technology officer for government innovation for the US government's Office of Science and Technology Policy. Her principal goal during her year at the White House was to create something more equivalent to the UK's Government Digital Service. She set in motion the creation of the United States Digital Service within the Executive Office of the President, and helped start 18F at the General Services Administration.
    </p>
    <p
      key="4"
    >
      Pahlka was awarded an Internet and Society Award from the Oxford Internet Institute, in recognition of her contribution to digital open government in the US.
    </p>
  </article>
  <section
    className="button-container"
  >
    <button
      className="button-back"
      onClick={[MockFunction]}
      value={0}
    >
      Go Back
    </button>
    <button
      className="button-quiz"
      onClick={[Function]}
    >
      Go to Quiz
    </button>
  </section>
</section>
`;

exports[`Article should match the snapshot with all data passed in 1`] = `
<section
  className="complete-article"
>
  <article
    className="pioneer-img-article"
  >
    <img
      className="pioneer"
      src="http://www.greatthoughtstreasury.com/sites/default/files/Jennifer-Pahlka1%5B1%5D.jpg"
    />
  </article>
  <article
    className="pioneer-article"
  >
    <p
      key="0"
    >
      Jennifer Pahlka (1969 - present day) is the founder and Executive Director of Code for America. She served as US Deputy Chief Technology Officer from June 2013 to June 2014 and helped found the United States Digital Service. Previously she had worked at CMP Media with various roles in the computer game industry. She was the co-chair and general manager of the Web 2.0 conferences.
    </p>
    <p
      key="1"
    >
      She was born in Port Deposit, Maryland, and raised in Austin, New Haven, and New York City. She is a graduate of the Bronx High School of Science and Yale University.
    </p>
    <p
      key="2"
    >
      Pahlka founded Code for America, a San Francisco-based non-profit organization that aims to make government more transparent and connected. According to the Washington Post Code for America is the technology world’s equivalent of the Peace Corps or Teach for America, offering an alternative to the old, broken path of government IT. In her 2012 TED Talk, Pahlka noted that we will not be able to reinvent government unless we also reinvent citizenship.
    </p>
    <p
      key="3"
    >
      In May 2013 Pahlka announced she was temporarily taking the position of deputy chief technology officer for government innovation for the US government's Office of Science and Technology Policy. Her principal goal during her year at the White House was to create something more equivalent to the UK's Government Digital Service. She set in motion the creation of the United States Digital Service within the Executive Office of the President, and helped start 18F at the General Services Administration.
    </p>
    <p
      key="4"
    >
      Pahlka was awarded an Internet and Society Award from the Oxford Internet Institute, in recognition of her contribution to digital open government in the US.
    </p>
  </article>
  <section
    className="button-container"
  >
    <button
      className="button-back"
      onClick={[MockFunction]}
      value={0}
    >
      Go Back
    </button>
    <button
      className="button-quiz"
      onClick={[Function]}
    >
      Go to Quiz
    </button>
  </section>
</section>
`;
